File: Algorithms/HeapSort.cpp
#include "HeapSort.h"
HeapSort::HeapSort() {

}
void HeapSort::Sort(int* arr,int count) {
	//Build a max heap for each sub tree of the binary tree
	for (int i = (count / 2) - 1; i > -1;i--) {
		this->Heapify(arr,count,i);
	}
	//Heapify the entire binary tree
	for (int i = 1; i < count;i++) {
		int unheapedIndex = count - i;
		this->Swap(arr,unheapedIndex,0);
		this->Heapify(arr,unheapedIndex,0);
	}
}
void HeapSort::Heapify(int* arr, int count,int rootIndex) {
	int largestIndex = rootIndex;
	int left = 2 * rootIndex + 1;
	int right = 2 * rootIndex + 2;

	if (left < count && arr[left] > arr[largestIndex]) {
		largestIndex = left;
	}
	if (right < count && arr[right] > arr[largestIndex]) {
		largestIndex = right;
	}

	//Swap if the largest number is not the root
	if (largestIndex != rootIndex) {
		this->Swap(arr, rootIndex, largestIndex);
		this->Heapify(arr,count,largestIndex);
	}
}
/// <summary>
/// Swap elements in the array (in place)
/// </summary>
/// <param name="arr">The array to swap elements</param>
/// <param name="f">first element</param>
/// <param name="s">second element</param>
void HeapSort::Swap(int*arr,int f,int s) {
	int temp = arr[f];
	arr[f] = arr[s];
	arr[s] = temp;
}

File: src/sort/heapSort.ts
import{Sort} from "./Sort";
export default class HeapSort extends Sort{
    sort(array: number[]): void | number[] {
        for(let i = (array.length/2) - 1; i > -1; i--){
            this.heapify(array,array.length,i);
        }

        for(let i = 1;i < array.length;i++){
            let unheapedIndex = array.length - i;
            HeapSort.swap(array,unheapedIndex,0);
            this.heapify(array,unheapedIndex,0);
        }
    }
    heapify(array:number[],count:number,index:number){
        let largestIndex = index;
        let compareIndex = 2 * index + 1;
        if(compareIndex < count && array[compareIndex] <array[largestIndex]){
            largestIndex = compareIndex;
        }
        compareIndex = 2 * index + 2;
        if(compareIndex < count && array[compareIndex] < array[largestIndex]){
            largestIndex = compareIndex;
        }

        if(largestIndex !== index){
            HeapSort.swap(array,largestIndex,index);
            this.heapify(array,count,largestIndex);
        }
    }
}